designs/helm-chart-autogenerate-plugin.md

```diff
@@ -0,0 +1,287 @@
+
+To achieve this goal, we are proposing a new native Kubebuilder [plugin](https://book.kubebuilder.io/plugins/plugins) (i.e., helm-chart/v1-alpha) which will provide the necessary scaffolds. The plugin will function similarly to the existing [Grafana Plugin](https://book.kubebuilder.io/plugins/grafana-v1-alpha), generating or regenerating HelmChart files using the init and edit sub-commands (i.e., `kubebuilder init|edit --plugins helm-chart/v1-alpha`). An alternative solution could be to implement an alpha command, similar to the [helper provided to upgrade projects](https://book.kubebuilder.io/reference/rescaffold).
+
+## Example
+
+**To enable the helm-chart generation when a project is initialized**
+
+> kubebuilder init --plugins=`go/v4,helm-chart/v1-alpha`
+
+**To enable the helm-chart generation after the project be scaffolded OR to update the helm-chart files**
+
+> kubebuilder edit --plugins=`go/v4,helm-chart/v1-alpha`
+
+#### Case 1: Initial Helm chart distribution
+
+For users who have developed an Kubebuilder Project and wish to distribute it via a Helm chart for the first time:
+
+By running `kubebuilder init|edit --plugins=helm-chart/v1alpha`, the scaffolds of Helm Chart to package the Project would be created in the project as follows:
+
+```shell
+$ tree
+.
+..
+├── PROJECT
+...
+├── deploy
+│   └── helm-chart // The helm chart with all values will be scaffold under a directory of the project
+...
+```
+
+The initial setup will be straightforward, creating the Helm chart structure and populating it with the default values as determined by the project's configuration.
+
+#### Case 2: Update the Helm chart with manifests changes
+
+For users who want to update an existing Helm chart to reflect changes in the kustomize directory, such as CRDs, Webhooks, or RBAC roles:
+
+The plugin will automates this process. When you run the plugin, it will generate the necessary manifests and synchronize them with the Helm chart. This ensures that all changes in your kubebuilder project files, such as CRDs, RBAC roles, and webhooks, are accurately reflected in the Helm chart.
+
+#### Case 3: Updating the Helm Chart with Customizations
+
+For users who have customized their Helm chart (e.g., added new configurations in `values.yaml`) and want to update it while preserving these customizations:
+
+Just like the other plugins in Kubebuilder, the Helm Chart plugin will not re-write the `values.yaml` after the initial generation. This means that any customizations made by the user will be preserved.
+
+## Open Questions [optional]
+
+- How to effectively manage the lifecycle of CRDs within the generated Helm chart to support upgrades and changes to CRDs while minimizing potential impacts on existing resources and user deployment processes?
+- How to deal with CertManager and Prometheus? Should we add them as dependencies to the Helm Chart?
+
```

Comment:
@dashanji @LCaparelli @Caiofralmeida, @jgillich, @jrxs-cg, @mhkarimi1383, @Jay-Madden, @maciaszczykm, @hprotzek, @octaviansima, @Cubixmeister, @isac322

Folks, 

I could generated a HelmChart using https://github.com/arttor/helmify.
So, I need your help to understand:
- Why kubebuilder maintainers would need to maintain a solution to package the its project in HelmChart when https://github.com/arttor/helmify is available already? Why not suggest it used instead?
What our plugin should do that is not done within? What value we bring for the community by doing this implementation instead of suggest its usage? What are the current problems faced with helmify that we would like to address here? Thank you for your help and collab.